gradlew.bat

```diff
@@ -33,7 +33,7 @@ set APP_HOME=%DIRNAME%
 for %%i in ("%APP_HOME%") do set APP_HOME=%%~fi
 
 @rem Add default JVM options here. You can also use JAVA_OPTS and GRADLE_OPTS to pass JVM options to this script.
-set DEFAULT_JVM_OPTS="-Xmx4G" "-Xms64m"
+set DEFAULT_JVM_OPTS="-Xmx64m" "-Xms64m"
```

Comment:
I highly disagree. For gradle runs you *need* a lot of memory otherwise the build will take for *ever*

Reply:
Wow. Yes, the automatic gradle self-update does _replace_ gradlew so we need to migrate individual settings. I didn't notice / remember we upped that. With your hint, I _do_ remember wondering how a ba can affect Linux results...

Upping gradle by just telling the wrapper to load a newer version by editing the config is not best practice, though it can be harmless for a time. Once in a while you should do a `./gradlew wrapper --gradle-version X` - add flavour X to taste. I just did that for the already "declared" gradle version, mainly so the wrapper _jar_ matches. Well, it also "upgrades" those scripts... Will patch soonish.

Reply:
- We already have `org.gradle.jvmargs=-Xmx4096m -XX:MaxMetaspaceSize=512m` in our root gradle.properties
- Rules for the following: DEFAULT_JVM_OPTS in both gradle scripts as per this PR, as placed by gradle wrapper update. Change jvmargs, invalidate cache and restart Studio, wait for indexing to finish. Gradle sync, then Build-clean and Build-rebuild. (Studio runs with -Xmx8g and actually uses <3g; system has 4g free and 13-14g assigned to cache over the entire test). 
- Rebuilding with `-Xmx6g -XX:MaxMetaspaceSize=512m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8` so I can be sure to _see_ these in system monitor
    - Sync took 49s
    - Rebuild took 5m14s
    - The daemon showed the -Xmx6g in its command line, but moved to a later part than the rest of org.gradle.jvmargs
    - The daemon spawns a sub-JVM for the actual compiler, it inherits -Xmx6g as can be seen in System Monitor
    - Daemon max actual Memory usage was 2.5G, the compiler around 1.5G, sum 4G - so -Xmx6g is way beyond saturation
- Switching to `-Xmx3072m`, clean, rebuild:
    - Sync took 53s
    - Rebuild took 46s - oops? 6x faster than 4g?
    - I get an additional daemon using -Xmx512m and not generating any disk read I/O
- Restart -Xmx3072m from 'invalidate cache'
    - Sync took 58s
    - Rebuild took 49s - oops? Despite invalidate and clean it still does not recompile and runs core:compileKotlin from cache instead.
- Restart -Xmx3072m, manual clean of ~/.gradle (yes it's outside the project what I need to clean, and yes it incurs 2G downloads, but they're outside measurement, the indexer already pulls them)
    - Sync took 53s
    - Rebuild took 5m16ss
- Switching to `-Xmx2g`
    - Sync same
    - 5m26s
- -Xmx1536m
    - Sync 9m23s (ignored as fluke)
    - Rebuild 4m57s
- -Xmx1g
    - Sync 1m5s
    - Rebuild 4m44s
    - Total processes 2.7g (gradle daemon 1.7, compiler 1)
- -Xmx768m
    - Sync 2m38s
    - Rebuild 5m1s
- -Xmx512m
    - Sync 2m51s
    - Rebuild: `java.lang.OutOfMemoryError: Java heap space` after 8m10s

- Conclusion: XX:MaxMetaspaceSize shows in memory usage; together with the observation that _two_ JVM's do the build, a -Xmx4g means a theoretical limit of ~10G memory usage. Our build already runs optimally with -Xmx2g, plus fuzzy factor -> 4g is perfect, maybe 3. The settings in our root gradle.properties _are_ respected, at least on Linux. I'd love to hear from a variety of platforms.

Reply:
> Rebuild took 46s - oops? 6x faster than 4g?

I would explain that as 3g were mapped fully in RAM, while 4g could use a swap file as a part of virtual memory, and of course, the file system is much slower than RAM.

Reply:
:upside_down_face:  On a 32G box that doesn't even bother assigning the remaining 3G free to file system cache as it seems to feel 13G are enough of that. Nah, swapping to SSD is probably not the explanation, some other app more likely (I know which one - it does massive io every few hours for a few minutes :cinema: ).